slack_bolt/slack_function/slack_function.py

```diff
@@ -0,0 +1,89 @@
+
+
+class SlackFunction:
+
+    func: Optional[Callable[..., Optional[BoltResponse]]] = None
+
+    def __init__(
+        self,
+        register_listener: Callable[..., Optional[Callable[..., Optional[BoltResponse]]]],
+        base_logger: Logger,
+        callback_id: Union[str, Pattern],
+    ):
+        self._register_listener = register_listener
+        self._base_logger = base_logger
+        self.callback_id = callback_id
+
+    def register_listener(
+        self,
+        functions: List[Callable[..., Optional[Callable[..., Optional[BoltResponse]]]]],
+        matchers: Optional[Sequence[Callable[..., bool]]] = None,
+        middleware: Optional[Sequence[Union[Callable, Middleware]]] = None,
+    ) -> Callable[..., Optional[Callable[..., Optional[BoltResponse]]]]:
+        primary_matcher = builtin_matchers.function_event(callback_id=self.callback_id, base_logger=self._base_logger)
+        self.func = self._register_listener(functions, primary_matcher, matchers, middleware, True)
+        return self
+
+    def __call__(self, *args, **kwargs) -> Optional[BoltResponse]:
+        if self.func is not None:
+            return self.func(*args, **kwargs)
+        return None
+
+    def action(
+        self,
+        constraints: Union[str, Pattern, Dict[str, Union[str, Pattern]]],
+        matchers: Optional[Sequence[Callable[..., bool]]] = None,
+        middleware: Optional[Sequence[Union[Callable, Middleware]]] = None,
+    ) -> Callable[..., Optional[Callable[..., Optional[BoltResponse]]]]:
+        """Registers a new action listener to your function. This method can be used as either a decorator or a method.
+
+            # Use this method as a decorator
+            @app.function("request-approval")
+            def request_approval(event, complete: Complete):
+                complete(outputs={})
+
+            @request_approval.action("approve_button")
+            def handle_request_approval_events(ack):
+                ack()
+
+            # Pass a function to this method
+            request_approval_func = app.function("request-approval")(request_approval)
+            request_approval.action("approve_button")(handle_request_approval_events)
+
+        * Refer to https://api.slack.com/reference/interaction-payloads/block-actions for actions in `blocks`.
+
+        To learn available arguments for middleware/listeners, see `slack_bolt.kwargs_injection.args`'s API document.
+
+        Args:
+            constraints: The conditions that match a request payload
+            matchers: A list of listener matcher functions.
+                Only when all the matchers return True, the listener function can be invoked.
+            middleware: A list of lister middleware functions.
+                Only when all the middleware call `next()` method, the listener function can be invoked.
+        """
+
+        middleware = list(middleware) if middleware else []
+        middleware.insert(0, AttachingFunctionToken())
```

Comment:
If you have a few listeners for function events, this middleware can be executed a few times. Can you move this to the global middleware list, not appending as a listener middleware to each listener?

Reply:
@seratch By adding the middleware globally won't this make it execute every time a new request comes in? My intended behavior was to have this middleware execute only when the request matches the `function` type listeners

Reply:
Maybe creating a middleware for a subgroup of listeners could be something worth exploring

Reply:
@WilliamBergamin 
> By adding the middleware globally won't this make it execute every time a new request comes in?

Yes, it can be but it's totally fine. If a request does not have the slack function token, the middleware does not do anything. The overhead is ignorable. More importantly, the middleware can be executed **only once** in any case. Contrarily, if we go with listener middleware for this and a developer has 10 functions in an app, the same middleware can be executed 10 times at maximum. This is because listener middleware can be executed every time bolt matches a request with a listener's matchers/middleware.

>My intended behavior was to have this middleware execute only when the request matches the function type listeners

In addition to above, if you have function listener at top of the list of middleware, the listener matcher can be executed for all incoming requests. Again, Bolt tries to find a listener to run for an incoming request. During this process, all the listener matchers/middleware can be executed for all requests.

So, in this case, we should go with a global middleware in the interest of simplicity and consistently small overhead.

Reply:
This has been implemented, thanks for explaining I initially thought this middleware was executed after the match occurred 💯